### PR TITLE
Clean up Compatible* usage

### PR DESCRIPTION
**Pull request by droark**
- Some work for conversion of v5 to v6 was put off earlier. Finish that work.
- Minor cleanup work discovered along the way.

### CODE REVIEW

**Comment by droark** [comment]
Leaving in draft form at the moment. I disabled some code in order to get the rest to compile. I need to go back and re-enable it, and double check some conversion output. After that, it should be good to go.

**Comment by droark** [comment]
@Quantumplation - Actually, I need to put this code back in draft form for a moment. Just realized a couple of entries currently marked as fake need to be updated. I'll have this ready in a minute.

**Comment by droark** [comment]
Ready to go.

File: ouroboros/chainsync/v5/types.go [hunk 1]
```diff
@@ -147,36 +270,84 @@ func (b BlockV5) ConvertToV6() chainsync.Block {
...
+		era = "byron"
+	} else if majorVer == 2 {
+		era = "shelley"
+	} else if majorVer == 3 {
+		era = "allegra"
+	} else if majorVer == 4 {
+		era = "mary"
+	} else if majorVer <= 6 {
+		era = "alonzo"
+	} else if majorVer == 7 {
```
Comment (Quantumplation) [inline ouroboros/chainsync/v5/types.go:339]: we can also proactively add `conway` as majorVer 8

File: ouroboros/chainsync/v5/types.go [hunk 2]
```diff
@@ -147,36 +270,84 @@ func (b BlockV5) ConvertToV6() chainsync.Block {
...
+		LeaderValue:            leaderValue,
+	}
+
+	// Unfortunately, due to how v5 data is formatted on the wire and stored by Ogmigo,
+	// we have to use other data to determine the era. Because the protocol version can
+	// be determined before a fork, unlike the slot number, we use the protocol version.
+	// This will make it easier to add support for future forks before the fork occurs,
+	// while also supporting testnet and other networks that use the same versioning.
+	var era string
+	if majorVer <= 1 {
```
[the remark this answers is not in the document]
Comment (Quantumplation) [reply ouroboros/chainsync/v5/types.go:329]: another way could be to define a constant map
```
const eras = map[int]string{
  0: "byron",
  1: "byron",
  2: "shelley",
  ...
}

...
Era: eras[majorVer]
```

File: ouroboros/chainsync/v5/types.go [hunk 3]
```diff
@@ -147,36 +270,84 @@ func (b BlockV5) ConvertToV6() chainsync.Block {
 		txArray = append(txArray, t.ConvertToV6())
 	}
 
-	fakeNonce := chainsync.Nonce{Output: "fake", Proof: "fake"}
+	nonceOutput := b.Header.Nonce["output"]
+	nonceProof := b.Header.Nonce["output"]
+	var nonce *chainsync.Nonce
+	if nonceOutput != "" && nonceProof != "" {
```
[the remark this answers is not in the document]
Comment (droark) [reply ouroboros/chainsync/v5/types.go:276]: I don't know if it happens. I'd have to go through all the data on the chain. That said, the spec seems to indicate that both fields are optional for both v5 and v6. I can change it such that we write if either one isn't empty.

File: ouroboros/chainsync/v5/types.go [hunk 4]
```diff
@@ -81,34 +100,138 @@ func (t TxV5) ConvertToV6() chainsync.Tx {
...
+// There's at least one special case in v6 (Mint) where the v5 "Coins" field doesn't
+// appear to be used. Give callers the option to skip it.
+func (v ValueV5) ConvertToV6(includeCoins bool) shared.Value {
+	assets := make(shared.Value)
+	adaMap := make(map[string]num.Int)
+	if includeCoins {
+		adaMap[shared.AdaAsset] = v.Coins
+		assets[shared.AdaPolicy] = adaMap
+	}
+	for policy, assetNum := range v.Assets {
```
[the remark this answers is not in the document]
Comment (droark) [reply ouroboros/chainsync/v5/types.go:217]: Thanks. I was just going off what I saw in the v5 spec, which, as you mentioned, used `policyId.assetName`.

File: ouroboros/chainsync/v5/types.go [hunk 5]
```diff
@@ -81,34 +100,138 @@ func (t TxV5) ConvertToV6() chainsync.Tx {
...
+	Coins  num.Int                    `json:"coins,omitempty"  dynamodbav:"coins,omitempty"`
+	Assets map[shared.AssetID]num.Int `json:"assets,omitempty" dynamodbav:"assets,omitempty"`
+}
+
+// There's at least one special case in v6 (Mint) where the v5 "Coins" field doesn't
+// appear to be used. Give callers the option to skip it.
+func (v ValueV5) ConvertToV6(includeCoins bool) shared.Value {
+	assets := make(shared.Value)
+	adaMap := make(map[string]num.Int)
+	if includeCoins {
```
[the remark this answers is not in the document]
Comment (droark) [reply ouroboros/chainsync/v5/types.go:213]: Just FYI, `if v.Coins != 0` doesn't work; `v.Coins` is `num.Int`. But, it's easy enough to do a workaround.

File: ouroboros/chainsync/v5/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Quantumplation) [reply ouroboros/chainsync/v5/types.go:217]: sure; what i mean is that `policy` is not an appropriate variable name; it's not a policy, it's an `assetId`, which is the combination of a policyId and assetName

File: ouroboros/chainsync/v5/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Quantumplation) [reply ouroboros/chainsync/v5/types.go:276]: nah, i was just curious

File: ouroboros/chainsync/v5/types.go [hunk 6]
```diff
@@ -147,36 +270,84 @@ func (b BlockV5) ConvertToV6() chainsync.Block {
...
-	fakeNonce := chainsync.Nonce{Output: "fake", Proof: "fake"}
+	nonceOutput := b.Header.Nonce["output"]
+	nonceProof := b.Header.Nonce["output"]
+	var nonce *chainsync.Nonce
+	if nonceOutput != "" && nonceProof != "" {
+		*nonce = chainsync.Nonce{Output: nonceOutput, Proof: nonceProof}
+	} else {
+		nonce = nil
+	}
+	majorVer := uint32(b.Header.ProtocolVersion["major"])
```
[the remark this answers is not in the document]
Comment (droark) [reply ouroboros/chainsync/v5/types.go:281]: I used it for the era check below. I figured I'd just read `b.Header.ProtocolVersion["major"]` once.

File: ouroboros/chainsync/v5/types.go [hunk 7]
```diff
@@ -147,36 +270,84 @@ func (b BlockV5) ConvertToV6() chainsync.Block {
...
-		if b.Header.OpCert["hotVk"].([]byte) != nil {
+		if b.Header.OpCert["hotVk"] != nil {
 			vk, _ = base64.StdEncoding.DecodeString(b.Header.OpCert["hotVk"].(string))
 		}
+		count := b.Header.OpCert["count"]
+		kesPd := b.Header.OpCert["kesPeriod"]
 		opCert = chainsync.OpCert{
-			Count: b.Header.OpCert["count"].(uint64),
-			Kes:   chainsync.Kes{Period: b.Header.OpCert["kesPeriod"].(uint64), VerificationKey: string(vk)},
+			Count: uint64(count.(float64)),
```
[the remark this answers is not in the document]
Comment (droark) [reply ouroboros/chainsync/v5/types.go:298]: Unforutnately, yes. [It's a JSON thing.](https://www.reddit.com/r/golang/comments/fhdvaj/uint64_via_json_becomes_different_value/fkahd42/) If I use anything else to cast `interface{}`, it chokes. 😞

File: ouroboros/chainsync/v5/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Quantumplation) [reply ouroboros/chainsync/v5/types.go:213]: `if v.Coins.Uint64() != 0` ;)

File: ouroboros/chainsync/v5/types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Quantumplation) [reply ouroboros/chainsync/v5/types.go:281]: oh ok, i skimmed and I didn't see it used again, that makes sense.

**Review by Quantumplation — APPROVED** [review]

**Pull request merged**